fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/JavaUdf.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * JavaUdf.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2023 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.query.plan.cascades.values;
+
+import com.apple.foundationdb.record.EvaluationContext;
+import com.apple.foundationdb.record.ObjectPlanHash;
+import com.apple.foundationdb.record.PlanHashable;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordStoreBase;
+import com.apple.foundationdb.record.query.plan.cascades.typing.Type;
+import com.google.common.collect.Streams;
+import com.google.protobuf.Message;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Represents a Java UDF.
+ */
+public abstract class JavaUdf extends AbstractValue {
+
+    @Nonnull
+    private static final ObjectPlanHash BASE_HASH = new ObjectPlanHash("Java-Udf-Value");
+
+    @Nonnull
+    final Iterable<? extends Value> children;
+
+    public JavaUdf(@Nonnull final Iterable<? extends Value> children) {
+        this.children = children;
+    }
+
+    @Nonnull
+    @Override
+    public Iterable<? extends Value> getChildren() {
+        return children;
+    }
+
+    @Nonnull
+    @Override
+    public Type getResultType() {
+        return getUdfResultType();
+    }
+
+    @Nonnull
+    public static Type getUdfResultType() {
+        return Type.primitiveType(Type.TypeCode.UNKNOWN);
+    }
+
+    @Nonnull
+    public static Iterable<Type> getUdfParameterTypes() {
+        return List.of();
```

Comment:
This should always return the declared types of the parameters I suppose. I think we should enforce the implementor to properly define those (same with the result type). On top of that the arity of this returned list (which should be a list not an iterable), should match the number of children I think. At least that would be a desirable invariant.